common/src/warp_runner/manager/commands/constellation_commands.rs

```diff
@@ -464,92 +472,112 @@ async fn handle_upload_progress(
     warp_storage: &mut warp_storage,
     mut upload_progress: ConstellationProgressStream,
     filename: String,
+    file_id: Uuid,
+    file_state: TransferState,
     file_path: PathBuf,
 ) {
     let tx_upload_file = UPLOAD_FILE_LISTENER.tx.clone();
     let mut previous_percentage: usize = 0;
     let mut upload_process_started = false;
+    let mut last_progress = None;
+    let mut paused = false;
 
-    while let Some(upload_progress) = upload_progress.next().await {
-        match upload_progress {
-            Progression::CurrentProgress {
-                name,
-                current,
-                total,
-            } => {
-                log::trace!("starting upload file action listener");
-                if let Ok(received_tx) = CANCEL_FILE_UPLOADLISTENER
-                    .rx
-                    .clone()
-                    .lock()
-                    .await
-                    .try_recv()
-                {
-                    if received_tx {
-                        let _ = tx_upload_file.send(UploadFileAction::Cancelling);
-                        break;
-                    }
+    loop {
+        tokio::select! {
+            biased;
+            true = file_state.matches(TransferStates::Cancel) => {
+                log::info!("{:?} file cancelled!", filename);
+                let _ = tx_upload_file.send(UploadFileAction::Cancelling(file_path, file_id));
+                return;
+            },
+            true = file_state.matches(TransferStates::Pause) => {
+                if !paused {
+                    let _ = tx_upload_file.send(UploadFileAction::Pausing(file_id));
+                    paused = true;
                 }
-                if !upload_process_started {
-                    upload_process_started = true;
-                    log::info!("Starting upload for {name}");
-                    log::info!("0% completed -> written 0 bytes")
+            },
+            upload_progress = upload_progress.next() => {
+                paused = false;
```

Comment:
We can check `paused` to determine if we can poll `upload_progress`

```suggestion
            upload_progress = upload_progress.next(), if !paused => {
```
Though note that we would need to have a previous condition to check to see if `TransferStates` is `TransferStates::Normal` so we can set that flag to false.

Reply:
hmm idk though.
not too much of a fan with this:
<img width="498" alt="Screenshot 2024-02-20 at 14 35 03" src="https://github.com/Satellite-im/Uplink/assets/34157027/55efcc7d-96af-4972-a8cb-c7605217614e">